modules/rostests/apitests/ntdll/RtlCriticalSection.c

```diff
@@ -0,0 +1,362 @@
+/*
+ * PROJECT:         ReactOS api tests
+ * LICENSE:         LGPLv2.1+ - See COPYING.LIB in the top level directory
+ * PURPOSE:         Test for Rtl Critical Section API
+ * PROGRAMMERS:     Thomas Faber <thomas.faber@reactos.org>
+ */
+
+#include "precomp.h"
+#include <pseh/pseh2.h>
+
+SYSTEM_INFO g_SysInfo;
+OSVERSIONINFOEXA g_VerInfo;
+ULONG g_Version;
+
+typedef
+NTSTATUS
+NTAPI
+FN_RtlInitializeCriticalSectionEx(
+    _Out_ PRTL_CRITICAL_SECTION CriticalSection,
+    _In_ ULONG SpinCount,
+    _In_ ULONG Flags);
+
+FN_RtlInitializeCriticalSectionEx* pfnRtlInitializeCriticalSectionEx;
+RTL_CRITICAL_SECTION CritSect;
+HANDLE hEventThread1Ready, hEventThread1Cont;
+HANDLE hEventThread2Ready, hEventThread2Cont;
+
+static
+void
+Test_Init(void)
+{
+    NTSTATUS Status;
+
+    ULONG DefaultSpinCount = (g_SysInfo.dwNumberOfProcessors > 1) ? 0x20007d0 : 0;
+    BOOL HasDebugInfo = (g_Version <= 0x600);
+
+    _SEH2_TRY
+    {
+        RtlInitializeCriticalSection(NULL);
+        Status = STATUS_SUCCESS;
+    }
+    _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+    {
+        Status = _SEH2_GetExceptionCode();
+    }
+    _SEH2_END;
+    ok_ntstatus(Status, STATUS_ACCESS_VIOLATION);
+
+    Status = RtlInitializeCriticalSection(&CritSect);
+    ok_ntstatus(Status, STATUS_SUCCESS);
+    ok_long(CritSect.LockCount, -1);
+    ok_long(CritSect.RecursionCount, 0);
+    ok_ptr(CritSect.OwningThread, NULL);
+    ok_ptr(CritSect.LockSemaphore, NULL);
+    ok_size_t(CritSect.SpinCount, DefaultSpinCount);
+    if (HasDebugInfo)
+    {
+        ok(CritSect.DebugInfo != NULL, "DebugInfo is %p\n", CritSect.DebugInfo);
+        ok(CritSect.DebugInfo != LongToPtr(-1), "DebugInfo is %p\n", CritSect.DebugInfo);
+    }
+    else
+    {
+        ok(CritSect.DebugInfo == LongToPtr(-1), "DebugInfo is %p\n", CritSect.DebugInfo);
+    }
+
+    Status = RtlInitializeCriticalSectionAndSpinCount(&CritSect, 0);
+    ok_ntstatus(Status, STATUS_SUCCESS);
+    ok_long(CritSect.LockCount, -1);
+    ok_long(CritSect.RecursionCount, 0);
+    ok_ptr(CritSect.OwningThread, NULL);
+    ok_ptr(CritSect.LockSemaphore, NULL);
+    ok_size_t(CritSect.SpinCount, DefaultSpinCount);
+    if (HasDebugInfo)
+    {
+        ok(CritSect.DebugInfo != NULL, "DebugInfo is %p\n", CritSect.DebugInfo);
+        ok(CritSect.DebugInfo != LongToPtr(-1), "DebugInfo is %p\n", CritSect.DebugInfo);
+    }
+    else
+    {
+        ok(CritSect.DebugInfo == LongToPtr(-1), "DebugInfo is %p\n", CritSect.DebugInfo);
+    }
+
+    Status = RtlInitializeCriticalSectionAndSpinCount(&CritSect, 0xFF000000);
+    ok_ntstatus(Status, STATUS_SUCCESS);
+    ok_size_t(CritSect.SpinCount, DefaultSpinCount);
+
+    if (pfnRtlInitializeCriticalSectionEx != NULL)
```

Comment:
What about:
```suggestion
    if (pfnRtlInitializeCriticalSectionEx == NULL)
    {
        skip("RtlInitializeCriticalSectionEx not available.\n");
        return;
    }
```

Reply:
No, because then someone adds tests at the end, and they will never run.

Reply:
yes maybe, if those other tests have nothing to do with the RtlInitialize***SectionEx function.